apps/rush-lib/src/logic/operations/OperationExecutionManager.ts

```diff
@@ -548,4 +588,131 @@ export class OperationExecutionManager {
+    const maxDurationLength: number = Math.max(
+      ...operations.map((operation) => operation.stopwatch.duration.toFixed(1).length + 1)
+    );
+    const chartWidth: number = TIMELINE_WIDTH - maxNameLength - maxDurationLength - 3;
+    //
+    // Loop through all operations, assembling some statistics about operations and
+    // phases, if applicable.
+    //
+
+    const output: string[] = [];
+
+    const durationByPhase: Map<string, number> = new Map();
+
+    const busyCpus: number[] = Array(this._parallelism).fill(-1);
+
+    output.push('='.repeat(TIMELINE_WIDTH));
+
+    for (const operation of operations) {
+      // Track time by phase
+      const phaseMatch: RegExpMatchArray | null = operation.name.match(/\((.+)\)$/);
+      if (phaseMatch) {
+        durationByPhase.set(
+          phaseMatch[1],
+          (durationByPhase.get(phaseMatch[1]) || 0) + operation.stopwatch.duration
+        );
+      }
+
+      // Track busy CPUs
+      const openCpu: number = busyCpus.findIndex((end) => end === -1 || end < operation.stopwatch.startTime!);
+      busyCpus[openCpu] = operation.stopwatch.endTime!;
+
+      // Build timeline chart
+
+      const startIdx: number = Math.floor(
+        ((operation.stopwatch.startTime! - allStart) * chartWidth) / allDuration
+      );
+      const endIdx: number = Math.floor(
+        ((operation.stopwatch.endTime! - allStart) * chartWidth) / allDuration
+      );
+      const length: number = endIdx - startIdx + 1;
+
+      const chart: string =
+        colors.gray('-'.repeat(startIdx)) +
+        TIMELINE_CHART_COLORIZER[operation.status](TIMELINE_CHART_SYMBOLS[operation.status].repeat(length)) +
+        colors.gray('-'.repeat(chartWidth - endIdx));
+      output.push(
+        colors.cyan(operation.name.padEnd(maxNameLength)) +
+          ' ' +
+          chart +
+          ' ' +
+          colors.white((operation.stopwatch.duration.toFixed(1) + 's').padStart(maxDurationLength))
+      );
+    }
+
+    output.push('='.repeat(TIMELINE_WIDTH));
+
+    //
+    // Format legend and summary areas
+    //
+
+    const usedCpus: number = busyCpus.filter((cpu) => cpu !== -1).length;
+
+    const legend: string[] = ['LEGEND:', '  [#] Success  [!] Failed/warnings  [%] Skipped/cached'];
+
+    const summary: string[] = [
+      'Total Work: ' + workDuration.toFixed(1) + 's',
+      'Wall Clock: ' + (allDuration / 1000).toFixed(1) + 's',
+      `Parallelism Used: ${usedCpus}/${this._parallelism}`
+    ];
+
+    output.push(legend[0] + summary[0].padStart(TIMELINE_WIDTH - legend[0].length));
+    output.push(legend[1] + summary[1].padStart(TIMELINE_WIDTH - legend[1].length));
+    output.push(summary[2].padStart(TIMELINE_WIDTH));
+
+    //
+    // Include time-by-phase, if phases are enabled
+    //
+
+    if (durationByPhase.size > 0) {
+      output.push('BY PHASE:');
+
+      const maxPhaseName: number = Math.max(16, ...[...durationByPhase.keys()].map((name) => name.length));
```

Comment:
Use a loop here. While strictly speaking having a few thousand phases would be unusual, it is allowed, and it's better not to error if that happens.